### ISSUE TITLE
producer ID request failure

### ISSUE THREAD

**Issue by g-freire** [twmb/franz-go#51]
Hi, I'm running franz at a production system. We are receiving the error from the image. The system looked okay in our tests, but after 7 days it triggered producer ID failure in one microservice and the same error happened in other microservice pod after 10 days. Unfortunally the verbose log was not activate, as we use a centralized log for all pods in production. Does anyone could give some light on it ? I'm concerned that this could be a non-deterministic issue, maybe cause by starvation or deadlocks and we like this lib very much.
Thanks
![image](https://user-images.githubusercontent.com/25310481/126655173-5eff3e53-f632-4a33-a1e8-efa318660c31.png)

**Comment by twmb** [comment]
What version of franz-go are you running? Under which behavior are you seeing that log? Does it look like things permanently fail, or do they recover? Are you using transactions?

**Comment by g-freire** [comment]
- What version of franz-go are you running? 
The version is  github.com/twmb/franz-go v0.7.5. 
- Under which behavior are you seeing that log? 
The behavior is unknow, because we can't reproduce it , even with load tests , seems that it happens after many days of usage.
- Does it look like things permanently fail, or do they recover? 
Looks like permanent fail, no recover.
- Are you using transactions?
We have the same problem with and without transactions.
- Also, do your clients wait a week before producing?
 No, the clients are actively producing, some times we have hours or days of inactivity, but most of times is being tested. 

Right now when we restart the pod everything is fine again. It affects some pods, not all of them, hence why we think it may be  some non-deterministic issue.

**Comment by twmb** [comment]
Is it possible to update to the latest version? I know there are some API changes but I hope those will be easy to fix and should make things better (I believe these are the last API changes, I'm intending to 1.0 when I have some time). There are some bugfixes, but I believe they're unrelated to what you're seeing.

The producer ID is only grabbed in a few places in the code (5 places, only once for non-transactional producing). It is possible that with transactional producing, this could leave the client in an unrecoverable state (because it does not currently think the error is recoverable).

Since this happens for you with and without transactions, I think it'll be easier if we focus on non-transactional right now -- transactions just increase the scope of the problem.

A producer ID is initialized once on the first produce, and can be failed and potentially reinitialized in certain error handling paths. For the non-transactional producer, the failure is preceeded by a log that looks like `batch errored with %s, failing the producer ID and resetting all sequence numbers`. Is it possible for you to confirm if you have ever seen this log?

Also, the non-transactional producer never goes through actually re-requesting the id, the non-transactional producer can bump its id locally. I'm not sure how you'd be seeing this error for non-transactional producers.

- What version of Kafka are your brokers on?
- Is it possible to get more debugging / logging information?
- Days of inactivity may be a problem from the Kafka side. Can you confirm whether or not this happens when clients are producing successfully for a while? Does this only happen after days of inactivity?
- What's your segment retention ms?

**Comment by g-freire** [comment]
Ok thanks again @twmb

- What version of Kafka are your brokers on?
2.7
- Is it possible to get more debugging / logging information?
Unfortunately we were not using the verbose log at the time and we may have lost it 
- Days of inactivity may be a problem from the Kafka side. Can you confirm whether or not this happens when clients are producing successfully for a while? Does this only happen after days of inactivity?
I will try to find this information. Our system is not using a continuous stream of data, we publish only after some database commands, which happens whenever we have a specific scheduled event or when testing.
- What's your segment retention ms?
![image](https://user-images.githubusercontent.com/25310481/126710516-ac65f829-ba7f-4a35-9c17-742b2b633933.png)

**Comment by twmb** [comment]
Looks like kowl, nice!

What is your `transactional.id.expiration.ms` setting? The default is 7 days I believe, so if you have not overridden it it will be 7 days.

I think we may be at a point of theorizing what the issue is without further logs -- are you willing to keep INFO logs for a few days in case this happens again?

**Comment by g-freire** [comment]
Yeah, we really dig kowl, very nice tool :)

I need to check the transactional.id.expiration.ms.
Yes, we will upgrade to the new version and run some pods with more verbose logs, if we get this error again I can open a new issue.

Thanks again @twmb for you time and quick responses. 
Have a great weekend. 
Cheers!

**Comment by twmb** [comment]
Thank you! I hope this was just a fluke, but if it happens again, I'm really keen to get to the bottom of this, so thank you for adding more logs!
